File: leetcode/151.reverse-words-in-a-string/go/reversewords.go
package reversewords

import "strings"

func reverseWords(s string) string {
	spaceRemoved := resolveSpace(s)
	reversedBytes := resolveLine(spaceRemoved)
	return resolveWords(reversedBytes)
}

func resolveSpace(s string) string {
	var sb strings.Builder
	sb.Grow(len(s))

	for i := 0; i < len(s); i++ {
		if s[i] == ' ' {
			continue
		}

		if sb.Len() > 0 && s[i-1] == ' ' {
			sb.WriteByte(' ')
		}
		sb.WriteByte(s[i])
	}

	return sb.String()
}

func resolveLine(s string) []byte {
	str := []byte(s)

	start := 0
	end := len(s) - 1

	for start < end {
		temp := str[start]
		str[start] = str[end]
		str[end] = temp

		start++
		end--
	}

	return str
}

func resolveWords(s []byte) string {
	start := 0

	for i := 0; i <= len(s); i++ {
		if i == len(s) || s[i] == ' ' {
			end := i - 1

			for start < end {
				temp := s[start]
				s[start] = s[end]
				s[end] = temp
				start++
				end--
			}
			start = i + 1
		}
	}

	return string(s)
}

File: leetcode/151.reverse-words-in-a-string/ts/reversewords.ts
const reverseWords = (s: string): string => {
    const spaceRemoved = resolveSpace(s)
    const reversedBytes = resolveLine(spaceRemoved)
    return resolveWords(reversedBytes)
};

const resolveSpace = (s: string): string[] => {
    const str: string[] = Array.from({ length: s.length })

    let idx = 0

    for (let i = 0; i < s.length; i++) {
        if (s[i] === ' ') continue

        if (idx > 0 && s[i - 1] === ' ') str[idx++] = ' '
        str[idx++] = s[i]
    }

    return str.slice(0, idx)
}

const resolveLine = (s: string[]): string[] => {
    let start = 0
    let end = s.length - 1

    while (start < end) {
        const temp = s[start]
        s[start] = s[end]
        s[end] = temp

        start++
        end--
    }

    return s
}

const resolveWords = (s: string[]): string => {
    let start = 0
    for (let i = 0; i <= s.length; i++) {
        if (i === s.length || s[i] === ' ') {
            let end = i - 1
            while (start < end) {
                const temp = s[start]
                s[start] = s[end]
                s[end] = temp

                start++
                end--
            }

            start = i + 1
        }
    }

    return s.join("")
}

export default reverseWords
